core/src/main/java/com/graphhopper/routing/weighting/custom/SpeedCustomConfig.java

```diff
@@ -0,0 +1,134 @@
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.routing.profiles.BooleanEncodedValue;
+import com.graphhopper.routing.profiles.DecimalEncodedValue;
+import com.graphhopper.routing.profiles.EncodedValueLookup;
+import com.graphhopper.routing.profiles.EnumEncodedValue;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.util.EdgeIteratorState;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.prep.PreparedGeometryFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static com.graphhopper.routing.weighting.custom.PriorityCustomConfig.createEnumToDoubleArray;
+import static com.graphhopper.routing.weighting.custom.PriorityCustomConfig.getEV;
+
+final class SpeedCustomConfig {
+    private List<EdgeToValueEntry> speedFactorList = new ArrayList<>();
+    private List<EdgeToValueEntry> maxSpeedList = new ArrayList<>();
+    private DecimalEncodedValue avgSpeedEnc;
+    private final double maxSpeed;
+    private final double maxSpeedFallback;
+
+    public SpeedCustomConfig(final double maxSpeed, CustomModel customModel, DecimalEncodedValue avgSpeedEnc,
+                             EncodedValueLookup lookup) {
+        this.maxSpeed = maxSpeed;
+        this.maxSpeedFallback = customModel.getMaxSpeedFallback() == null ? maxSpeed : customModel.getMaxSpeedFallback();
+        this.avgSpeedEnc = avgSpeedEnc;
+        if (this.maxSpeedFallback > maxSpeed)
+            throw new IllegalArgumentException("max_speed_fallback cannot be bigger than max_speed " + maxSpeed);
+
+        // use max_speed to lower speed for the specified conditions
+        for (Map.Entry<String, Object> entry : customModel.getMaxSpeed().entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+
+            if (value instanceof Number) {
+                if (key.startsWith(GeoToValueEntry.key(""))) {
+                    Geometry geometry = GeoToValueEntry.pickGeometry(customModel, key);
+                    maxSpeedList.add(new GeoToValueEntry(new PreparedGeometryFactory().create(geometry), ((Number) value).doubleValue(), maxSpeed));
+                } else {
+                    BooleanEncodedValue encodedValue = getEV(lookup, "max_speed", key, BooleanEncodedValue.class);
+                    maxSpeedList.add(new BooleanToValueEntry(encodedValue, ((Number) value).doubleValue(), maxSpeed));
```

Comment:
Do we not have to limit the max_speed values to the encoders max speed for boolean and geo, just like we do for enum? Same thing for speed_factor.

Reply:
I think we already do this. SpeedCustomConfig uses baseFlagEncoder.getMaxSpeed() as first argument => maxSpeed. Or did I misunderstand what you mean here?

Reply:
Ah at the very end `calcSpeed` limits the speed to `maxSpeed(=baseFlagEncoder.getMaxSpeed())`. I got confused, because compared to Geo- and BooleanToValueEntry, we restrict the maximum speed already when we create the `EnumToValueEntries` (in `createEnumToDoubleArray`). So I guess we get an error if a max speed is too high for enum encoded values, but for boolean and geo it will only be ignored.

Reply:
Have created a change where we raise an exception for geo and boolean too